src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1060,6 +1063,12 @@ CreateDistributedPlan(uint64 planId, Query *originalQuery, Query *query, ParamLi
 	 */
 	if (list_length(subPlanList) > 0 || hasCtes)
 	{
+		/*
+		 * we should keep track of all recursive subplans, otherwise multi level
+		 * recursive calls can overwrite subPlans for distributed plans.
+		 */
+		recursiveSubPlanList = list_concat(recursiveSubPlanList, subPlanList);
```

Comment:
(covered in offline discussion)

This is not the right fix, since both the first and the second call to GenerateSubplansForSubqueriesAndCTEs will number its subplans from 1, so we get different subplans with the same ID. A better fix is to makes sure recursive planning covers everything so we don't do recursive planning again the second time around.